packages/babel-parser/src/parser/statement.ts

```diff
@@ -522,13 +522,49 @@ export default abstract class StatementParser extends ExpressionParser {
     }
   }
 
+  decoratorsEnabledBeforeExport(): boolean {
+    if (this.hasPlugin("decorators-legacy")) return true;
+    return (
+      this.hasPlugin("decorators") &&
+      !!this.getPluginOption("decorators", "decoratorsBeforeExport")
+    );
+  }
+
   takeDecorators(node: N.HasDecorators): void {
     const decorators =
       this.state.decoratorStack[this.state.decoratorStack.length - 1];
     if (decorators.length) {
       node.decorators = decorators;
       this.resetStartLocationFromNode(node, decorators[0]);
       this.state.decoratorStack[this.state.decoratorStack.length - 1] = [];
+
+      // In case of decorators followed by `export`, we manually mark the
+      // comments between `export` and `class` as leading comments of the
+      // class declaration node.
+      // This is needed because the class location range overlaps with
+      // `export`, and thus they would me marked as inner comments of
+      // the class declaration.
+      //   @dec export /* comment */ class Foo {}
+      // See [parseDecorators] for the hanling of comments before decorators
+      if (this.decoratorsEnabledBeforeExport()) {
+        const {
+          start,
+          lastTokStart,
+          lastTokEndLoc: { index: lastTokEnd },
+        } = this.state;
+
+        if (this.input.slice(lastTokStart, lastTokEnd) === "export") {
```

Comment:
Can we move the comments manipulation to https://github.com/babel/babel/blob/ddaee9025c3af7d7eebdda62c89784074c8c034a/packages/babel-parser/src/parser/comments.ts#L164 ?

So that comments manipulation are placed within a single file and we probably don't have to do an `export` lookback.

Reply:
I moved the logic to comments.ts, but I'm not sure about how to do it in finalizeComment. Specifically, I don't think we can distinguish `/* 1 */` and `/* 2 */` when finalizing the comments of the class node:
```js
@dec export default /* 1 */ class /* 2 */ extends X {}
```

(I just realized that this PR doesn't support the `default` case, I'll add it)